BlueprintUI/Tests/GridRowTests.swift

```diff
@@ -0,0 +1,228 @@
+        let gridRow = GridRow()
+        XCTAssertTrue(gridRow.children.isEmpty)
+        XCTAssertEqual(gridRow.verticalAlignment, .fill)
+        XCTAssertEqual(gridRow.spacing, 0)
+    }
+
+    func test_measure_unconstrained() {
+        do {
+            // #1: Proportional child measurment
+            //  proportions:      1,     2
+            //  measured widths:  10,    25
+            //  scales:           10,    12.5
+            //  max scale:        12.5
+            //  width:            (1 * 12.5) + (2 * 12.5) = 37.5
+            //  height:           max(5, 10) = 10
+            let gridRow = GridRow { row in
+                row.add(child: .proportional(1, TestElement(size: CGSize(width: 10, height: 5))))
+                row.add(child: .proportional(2, TestElement(size: CGSize(width: 25, height: 10))))
+            }
+            XCTAssertEqual(gridRow.content.measure(in: .unconstrained), CGSize(width: 37.5, height: 10))
+        }
+
+        do {
+            // #2: Absolute and proportional child layout
+            //  Same approach as #1, plus a 1x8 absolutely-sized child. Note the absolute sizing stretched this
+            //  child to a width of 30.
+            //   width:   37.5 + 30 = 67.5
+            //   height:  max(5, 10, 50) = 50
+            let gridRow = GridRow { row in
+                row.add(child: .proportional(1, TestElement(size: CGSize(width: 10, height: 5))))
+                row.add(child: .proportional(2, TestElement(size: CGSize(width: 25, height: 10))))
+                row.add(child: .absolute(30, TestElement(size: CGSize(width: 1, height: 50))))
+            }
+            XCTAssertEqual(gridRow.content.measure(in: .unconstrained), CGSize(width: 67.5, height: 50))
+        }
+
+        do {
+            // #3: Spacing
+            //  Same approach as #1, plus a 5 point spacing between elements.
+            //  width:   37.5 + 5.0 = 42.5
+            //  height:  10
+            let gridRow = GridRow { row in
+                row.spacing = 5
+                row.add(child: .proportional(1, TestElement(size: CGSize(width: 10, height: 5))))
+                row.add(child: .proportional(2, TestElement(size: CGSize(width: 25, height: 10))))
+            }
+            XCTAssertEqual(gridRow.content.measure(in: .unconstrained), CGSize(width: 42.5, height: 10))
+        }
+    }
+
+    func test_measure_constrained() {
+        //  Proportional children size to fit the available width. The tallest child is used as the measured height.
+        let gridRow = GridRow { row in
+            row.add(child: .proportional(1, TestElement(size: CGSize(width: 10, height: 5))))
+            row.add(child: .proportional(2, TestElement(size: CGSize(width: 25, height: 10))))
+        }
+        let constraint = SizeConstraint(width: .atMost(25), height: .atMost(100))
+        XCTAssertEqual(gridRow.content.measure(in: constraint), CGSize(width: 25.0, height: 10))
+    }
+
+    func test_layout() {
```

Comment:
Good coverage! Here's a couple more corner cases that might be useful:
- a row with only absolute children
- a child with proportion of zero
- a row that is too large for the available size (OK to have overflowing frames, but the proportional ones shouldn't be negative)

Reply:
Added a `test_layout_edgeCases()` method with these tests. Thanks for the suggestion.